rfcs/0003-gene-sets/api.md

```diff
@@ -0,0 +1,463 @@
+
+- [US1] A user should be able to export a gene set, if the gene set was created by differential expression, the pvalue, log fold change, and category label fields will be populated in addition to the gene name. This will be available in a csv file (see data formats for more detail).
+
+### Users can upload their own gene sets and see them in the right side bar with the full range of current cellxgene functionalities
+
+- [US2] A user should be able to upload a gene set via a csv file (see data formats for more detail) for a dataset.
+- [US3] A user should be able to upload a gene set they previously downloaded from cellxgene
+- [US4] When a user uploads a gene set, that list of genes should appear in the right side bar.
+- [US5] If the user was logged in when they uploaded a gene set that gene set should persist if they logout and back in or reload the page
+- [US6] If the gene names match those used in the dataset the user should be able to use general cellxgene functionality
+  (color by genes, calculate differential expression levels, etc.) on those genes.
+- [US7] If a gene in the user uploaded gene set does not exist in the dataset the user will see an error message specifying the gene that wasnt in the dataset, but the remaining genes will be displayed.
+
+### If there is a standard format for differential expression output, we should follow that. If not, the format should be as straightforward as possible while serving these above cases
+
+### Nonfunctional requirements
+
+- Users should be able to export precomputed differential expression results in real time
+
+## Detailed Design | Architecture | Implementation
+
+Because gene sets are closely tied to differential expression, some aspects of the differential expression design plans will also be detailed below.
+
+### User uploaded gene sets
+
+#### 1. User creates csv containing a list of genes (and optional comments about why each gene was included)
+
+CSV must follow format described in the Data Model section
+
+#### 2. User uploads file to hosted cellxgene client
+
+The user will typically upload a file from their local machine but we should consider supporting uploads from s3 in future iterations.
+
+The user must be logged in to upload a gene set.
+
+Users will upload the file by passing its path to the client.
+The client will validate the format of the csv and check that the genes listed in the csv exist in the dataset
+
+- if the format is invalid the client will raise a descriptive error message to the user
+- if the file is valid, the client will display the genes in the right side bar
+- if the file is valid, but it references genes that are not in the dataset, the client will display the genes that do
+  exist in the right side bar and alert the user that their var set contained genes that are not a part of the dataset and they have
+  been removed from the var set that will be stored/displayed.
```

Comment:
Will you make a distinction between genes that don't exist in the entire AnnData object and genes that are in the raw data, but missing from additional layers (normalized/subsetted versions) of the data?